middleware.go

```diff
@@ -0,0 +1,640 @@
+		if err != nil {
+			return []ProcessedRecord{ErrorRecord{Error: fmt.Errorf("record %d: %w", i, err)}}
+		}
+		payloadSubjectVersions[i] = sv
+	}
+
+	// process records
+	recsOut := p.Processor.Process(ctx, records)
+
+	// encode key and payload with schema if subject and version didn't change
+	for i, rec := range recsOut {
+		if i >= len(keySubjectVersions) {
+			break // should not happen, if the processor respects the input length
+		}
+
+		singleRec, ok := rec.(SingleRecord)
+		if !ok {
+			continue
+		}
+		if err := p.encodeKey(ctx, (*opencdc.Record)(&singleRec), keySubjectVersions[i]); err != nil {
+			recsOut[i] = ErrorRecord{Error: err}
+		}
+		if err := p.encodePayload(ctx, (*opencdc.Record)(&singleRec), payloadSubjectVersions[i]); err != nil {
+			recsOut[i] = ErrorRecord{Error: err}
+		}
+		recsOut[i] = singleRec
+	}
+	return recsOut
+}
+
+func (p *processorWithSchemaEncode) keySubjectVersion(rec opencdc.Record) (subjectVersion, error) {
+	subject, err := rec.Metadata.GetKeySchemaSubject()
+	if err != nil && !errors.Is(err, opencdc.ErrMetadataFieldNotFound) {
+		return subjectVersion{}, fmt.Errorf("failed to get key schema subject: %w", err)
+	}
+	version, err := rec.Metadata.GetKeySchemaVersion()
+	if err != nil && !errors.Is(err, opencdc.ErrMetadataFieldNotFound) {
+		return subjectVersion{}, fmt.Errorf("failed to get key schema version: %w", err)
+	}
+	return subjectVersion{subject: subject, version: version}, nil
+}
+
+func (p *processorWithSchemaEncode) payloadSubjectVersion(rec opencdc.Record) (subjectVersion, error) {
+	subject, err := rec.Metadata.GetPayloadSchemaSubject()
+	if err != nil && !errors.Is(err, opencdc.ErrMetadataFieldNotFound) {
+		return subjectVersion{}, fmt.Errorf("failed to get payload schema subject: %w", err)
+	}
+	version, err := rec.Metadata.GetPayloadSchemaVersion()
+	if err != nil && !errors.Is(err, opencdc.ErrMetadataFieldNotFound) {
+		return subjectVersion{}, fmt.Errorf("failed to get payload schema version: %w", err)
+	}
+	return subjectVersion{subject: subject, version: version}, nil
+}
+
+func (p *processorWithSchemaEncode) encodeKey(ctx context.Context, rec *opencdc.Record, incomingSubjectVersion subjectVersion) error {
+	if !p.keyEnabled {
+		return nil // key schema encoding is disabled
+	}
+	if _, ok := rec.Key.(opencdc.StructuredData); !ok {
+		// log warning once, to avoid spamming the logs
+		p.keyWarnOnce.Do(func() {
+			Logger(ctx).Warn().Msgf(`record key is not structured, consider disabling the processor schema key encoding using "%s: false"`, configProcessorSchemaEncodeKeyEnabled)
+		})
+		return nil
+	}
+
+	sv, err := p.keySubjectVersion(*rec)
+	if err != nil {
+		return err // already wrapped
+	}
+	oldSchema, err := sdkschema.Get(ctx, sv.subject, sv.version)
+	if err != nil {
+		return fmt.Errorf("failed to get schema for key: %w", err)
+	}
+	if sv != incomingSubjectVersion {
+		return nil // key schema subject or version changed, skip encoding
```

Comment:
Why do you think we should skip encoding?

Reply:
Good thing you brought it up, that's a behavior that's debatable.

I thought about when this middleware should kick in. Ideally, it should be used together with the decoding middleware, so the decoding one decodes Avro to structured data, the processor changes the structured data, and then this middleware extracts the schema and encodes it back into Avro.

Now, what if there was no schema attached at the start? Then the decoding middleware wouldn't decode anything, and it would be a bit unexpected if this middleware kicked in to encode it to Avro, since it's not necessarily what the incoming record contained.

And what if the middleware changes the subject and/or version? At that point I'm not exactly sure what would be the expected behavior. If they were removed, the processor might not want to use Avro anymore. If they were changed, the processor has maybe encoded the data manually and used a different schema that it has used to overwrite the metadata.

That's at least my reasoning, I'm interested in your view though.

Reply:
> If they were changed, the processor has maybe encoded the data manually and used a different schema that it has used to overwrite the metadata.

I think the keyword is "maybe". There's a chance that wasn't done and that the schema subject/version pair doesn't match the key/payload anymore.

Downstream, in a destination connector, that same key/payload will be decoded using the new subject/version. So if the middleware or the processor itself didn't make sure that the key/payload are correctly encoded, we'll get an error. And I believe either failing sooner (in the processor is better) or re-encoding is better. Re-encoding is an additional performance hit, especially if the processor already did it.

All in all, I think it's around making sure that a key/payload was actually encoded with the subject/version that's in the metadata. We can also say that it's not our responsibility to validate the data.

Reply:
So you're saying that if the subject/version changed we should check if we still have structured data, if we do, we take the attached schema from the metadata and encode it? That sounds reasonable, similar to what the source middleware is doing. I'll implement it 👍

Reply:
> what if there was no schema attached at the start? Then the decoding middleware wouldn't decode anything, and it would be a bit unexpected if this middleware kicked in to encode it to Avro

I'll change this behavior so that it _will_ be encoded, as long as the outgoing record has a schema attached in the metadata.